Refactoring of ParticleTracker, Plasma class

StanczakDominik opened the pull request:
Closes #636 

This is going to be a pretty open-ended one. I've got a few ideas I'd like to tinker with:

- [x] Remove field interpolation code from ParticleTracker, as right now it's completely tied to Plasma3d on a uniform grid. That's completely idiotic and [those responsible should be sacked](https://github.com/PlasmaPy/PlasmaPy/blame/master/plasmapy/simulation/particletracker.py).
- [x] Put field interpolation in Plasma3d, because that's reasonable.
- [x] Replace `plasma3d` with draft AnalyticalPlasma in ParticleTracker, which probably shares a whole bunch of functionality with PlasmaBlob
- [x] reenable some of the tests
- [x] optimize ParticleTracker
- [ ] merge PlasmaBlob with AnalyticalPlasma as needed

Leftovers for future work:
   
- [x] Experiment with xarray for Plasma3D - working with the bare numpy arrays is proving to be a pain.
     - [ ] test field interpolation in there.
- [ ] Use notes from #569

pep8speaks commented:
Hello @StanczakDominik! Thanks for updating your pull request.



* In the file [`plasmapy/classes/sources/analyticalplasma.py`](https://github.com/PlasmaPy/PlasmaPy/blob/1ba55df98a08fc5106fb1be9af458243fbc1af78/plasmapy/classes/sources/analyticalplasma.py):

> [Line 12:41](https://github.com/PlasmaPy/PlasmaPy/blob/1ba55df98a08fc5106fb1be9af458243fbc1af78/plasmapy/classes/sources/analyticalplasma.py#L12): [E124](https://duckduckgo.com/?q=pep8%20E124) closing bracket does not match visual indentation





* In the file [`plasmapy/simulation/tests/test_particletracker.py`](https://github.com/PlasmaPy/PlasmaPy/blob/1ba55df98a08fc5106fb1be9af458243fbc1af78/plasmapy/simulation/tests/test_particletracker.py):

> [Line 40:5](https://github.com/PlasmaPy/PlasmaPy/blob/1ba55df98a08fc5106fb1be9af458243fbc1af78/plasmapy/simulation/tests/test_particletracker.py#L40): [E306](https://duckduckgo.com/?q=pep8%20E306) expected 1 blank line before a nested definition, found 0
> [Line 46:17](https://github.com/PlasmaPy/PlasmaPy/blob/1ba55df98a08fc5106fb1be9af458243fbc1af78/plasmapy/simulation/tests/test_particletracker.py#L46): [E128](https://duckduckgo.com/?q=pep8%20E128) continuation line under-indented for visual indent
> [Line 107:5](https://github.com/PlasmaPy/PlasmaPy/blob/1ba55df98a08fc5106fb1be9af458243fbc1af78/plasmapy/simulation/tests/test_particletracker.py#L107): [E306](https://duckduckgo.com/?q=pep8%20E306) expected 1 blank line before a nested definition, found 0
> [Line 109:18](https://github.com/PlasmaPy/PlasmaPy/blob/1ba55df98a08fc5106fb1be9af458243fbc1af78/plasmapy/simulation/tests/test_particletracker.py#L109): [E222](https://duckduckgo.com/?q=pep8%20E222) multiple spaces after operator
> [Line 148:5](https://github.com/PlasmaPy/PlasmaPy/blob/1ba55df98a08fc5106fb1be9af458243fbc1af78/plasmapy/simulation/tests/test_particletracker.py#L148): [E306](https://duckduckgo.com/?q=pep8%20E306) expected 1 blank line before a nested definition, found 0
> [Line 151:31](https://github.com/PlasmaPy/PlasmaPy/blob/1ba55df98a08fc5106fb1be9af458243fbc1af78/plasmapy/simulation/tests/test_particletracker.py#L151): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ','
> [Line 154:18](https://github.com/PlasmaPy/PlasmaPy/blob/1ba55df98a08fc5106fb1be9af458243fbc1af78/plasmapy/simulation/tests/test_particletracker.py#L154): [E222](https://duckduckgo.com/?q=pep8%20E222) multiple spaces after operator

StanczakDominik commented on plasmapy/simulation/particletracker.py:
```diff
@@ -156,36 +182,51 @@ def boris_push(self, init=False):
+    @numba.njit(parallel=True)
+    # @profile
+    def _boris_push(x, v, b, e, hqmdt, dt):
+        for i in numba.prange(len(x)):
```
this is thus parallelized over the number of particles and goes pretty darn fast!

StanczakDominik commented on plasmapy/simulation/particletracker.py:
```diff
@@ -248,7 +289,20 @@ def plot_time_trajectories(self, plot="xyz"):  # coverage: ignore
-            "Kinetic energy is not conserved!"
+        conservation = np.allclose(self.kinetic_energy_history,
+                                   self.kinetic_energy_history.mean(),
+                                   atol=3 * self.N * self.kinetic_energy_history.std())
```
Okay, I'm an idiot, I increased this tolerance by the N factor because I thought this was some issue with running multiple particles.

BUT IT FAILS FOR A NONZERO ELECTRIC FIELD!

Needs reverting.

StanczakDominik commented on plasmapy/simulation/tests/test_particletracker.py:
```diff
@@ -41,64 +31,79 @@ def sine(t, amplitude, omega, phase, mean):
-#         Tests the particle stepper for a uniform magnetic field motion.
-#     """
-#     test_plasma = uniform_magnetic_field()
+E_unit = u.V / u.m
```
Precomputing this as per https://docs.astropy.org/en/stable/units/#performance-tips

StanczakDominik replied:
The performance bump was pretty horrible. It made interpolating fields the slowest part, actually.

StanczakDominik commented on plasmapy/simulation/tests/test_particletracker.py:
```diff
@@ -41,64 +31,79 @@ def sine(t, amplitude, omega, phase, mean):
+    r"""
+        Tests the particle stepper for a uniform magnetic field motion.
+    """
+    # @profile
```
Obvious cleanup required

StanczakDominik commented on plasmapy/simulation/tests/test_particletracker.py:
```diff
@@ -132,75 +140,83 @@ def test_particle_exb_drift(uniform_magnetic_field):
 
 
-""" TODO: figure out how to get this test to work
+@pytest.mark.skip
```
I'll be trying to do something interesting with this one, but I might just use a more interesting field and stick that into the example.

StanczakDominik commented on plasmapy/simulation/tests/test_particletracker.py:
```diff
@@ -132,75 +140,83 @@ def test_particle_exb_drift(uniform_magnetic_field):
-#     ParticleTracker(plasma, 'e', dt=1e-14*u.s, nt=2).run()
+if __name__ == "__main__":
+    import pytest
+    pytest.main([__file__, "-s"])
```
This was helpful while profiling https://stackoverflow.com/a/23765051/4417567

StanczakDominik commented on tox.ini:
```diff
@@ -16,6 +16,8 @@ deps =
     pytest-cov
     pytest-xdist
+    pudb
+    pytest-pudb
```
Temporary fix for some issues on my setup, this will probably not be necessary (though I wouldn't mind keeping it in if we officially sanction pudb as the most awestome debugger there is!)

StanczakDominik commented on plasmapy/simulation/tests/test_particletracker.py:
```diff
@@ -41,64 +31,79 @@ def sine(t, amplitude, omega, phase, mean):
-#                       atol=estimated_gyrofrequency_std), \
-#         "Gyrofrequencies don't match!"
+    s = ParticleTracker(test_plasma, particle_type=particle_type, dt=dt, nt=int(1e4))
+    s._v[:, 1] = perp_speed.si.value
```
I had some issues with directly modifying `s.v[:, 1] = perp_speed`, still figuring that one out

StanczakDominik replied:
oh crud, https://stackoverflow.com/questions/3137685/using-property-decorator-on-dicts I think this is gonna be basically impossible

namurphy commented on plasmapy/simulation/particletracker.py:
```diff
@@ -75,43 +85,57 @@ def __init__(self, plasma, particle_type='p', n_particles=1, scaling=1,
+                                          dtype=float)
+        self._velocity_history = np.zeros((self.NT, *self.v.shape),
+                                          dtype=float)
+        self._hqmdt = (self.eff_q / self.eff_m / 2 * dt).si.value
```
I am wondering if there is a better variable name than `hqmdt`.

StanczakDominik replied:
There most certainly is. But if I wrote "half charge timestep over mass", line length would explode in a few places.

On the other hand, I have been reading Computer Simulation Using Particles recently, so my thinking may be contaminated by the FORTRAN examples...

StanczakDominik commented on plasmapy/classes/sources/plasma3d.py:
```diff
@@ -7,6 +7,7 @@
 import itertools
 
 from plasmapy.physics.magnetostatics import MagnetoStatics
+import scipy.interpolate as interp
```
I think there might already be one issue like that!

tag: open research question, maybe?

StanczakDominik commented on plasmapy/classes/sources/plasma3d.py:
```diff
@@ -115,6 +140,7 @@ def is_datasource_for(cls, **kwargs):
             match = False
         return match
 
+    # TODO this needs more numba and less for loop
```
This is now #684

StanczakDominik commented:
I'm giving up on running calculations on `Quantities`. We'll have to live with setting velocities and positions via `_x`, `_v` for now.

StanczakDominik commented:
This is about done; what's left is:

- [ ] writing tests and the basics for `AnalyticalPlasma`
- [ ] ... if we can't merge `AnalyticalPlasma` and `PlasmaBlob`

StanczakDominik commented:
I'm having a few doubts about the design here that I'll have to write down when I can muster some energy for it, right now, it's exhausting...

StanczakDominik commented:
Progress! I should have something to share here soon-ish! :)

StanczakDominik commented:
I wrote some basic viz for trajectories using mayavi, but that lib is proving to be difficult to actually install via CI. Ideally I'd be able to at least run readthedocs with it; that way we can show the visualization off in docs.

I was only able to get things working decently quickly after adding non quantity versions for most of the stuff I needed. This is not beautiful by any means, but it gets the job done. 

I also added the positions and velocities as private variables and provide access to the quantity ones via properties. This is because having those as the base data structures lets us do access faster than accessing quantity.si.value. I don't particularly enjoy doing it this way, but it is what it is.

I also commandeered the plasma class - brutally, at that - to add two new cases. AnalyticalPlasma, which was there in the first "sprint", takes functions of position for electric and magnetic fields. This lets you easily provide those to look at, say, particle drifts. I wanted these to generalize to multiparticle simulations, so they have a few hacks in there to ensure that can be done. This might be better renamed as "AnalyticalFields"...

I also added a Coils object that provides the same field interface and is internalized by a number of MagnetoStatics. This lets you use a few of those to construct a very basic model of a tokamak. Unfortunately to make this go sufficiently fast I had to do a few dirty tricks that result from numba jitting methods not really working right now - self does not compile to a basic numeric type, so I can't access class fields as parameters.

I also think it might make sense to have a CompositePlasma object that's the result of adding, say, AnalyticalFields and Coils - and then you could just iterate over its components in applications. Future work, though.

StanczakDominik commented:
Things still to do (future issues):

# rewrite `ParticleStepperSolution` using `xarray`

I think it's going to be a particularly good fit for that one and should make access much more comfortable.

# Adaptive time stepper

This is going to be a bit more complex:

* Position, velocity and time histories are going to have to become lists collected at each step. That's probably a good idea anyway, as with xarray + potentialy zarr support, we'll be able to run longer simulations by clearing out the history lists once they grow over some preset conditions and dump them to xarray/zarr/disk memory.
	* then we can also simplify mayavi plotting in terms of memory via sampling every N-th point.
* We'll need a clever criterion for actually changing the 

# Bring other magnetostatics in line with GeneralWire, CircularWire

I added `_magnetic_field` (the unit-less version) to those two, but didn't follow up with the rest. Thus, they're incompatible with ParticleTracker for now.

# Magnetostatics have no units in reprs

StanczakDominik commented:
Turns out I have no idea what's installing `hypothesis` the conda test envs, so I'm just xfailing the hypothesis test I added.

namurphy commented on plasmapy/classes/sources/coils.py:
```diff
@@ -0,0 +1,83 @@
+        return fig
+
+    @classmethod
+    def toykamak(
```
Should this be tokamak, or does it refer to a toy model of a tokamak?

StanczakDominik replied:
Yes. :)

namurphy commented on plasmapy/simulation/particletracker.py:
```diff
@@ -1,203 +1,134 @@
-    """
-    Object representing a species of particles: ions, electrons, or simply
-    a group of particles with a particular initial velocity distribution.
+@numba.njit(parallel = True)
```
Yay Numba!

namurphy commented on plasmapy/simulation/particletracker.py:
```diff
@@ -1,203 +1,134 @@
+    References
+    ----------
+    .. [1] C. K. Birdsall, A. B. Langdon, "Plasma Physics via Computer
+           Simulation", 2004, p. 58-63
```
I need to start reading this book...

StanczakDominik replied:
I need to finish it...

namurphy commented on plasmapy/simulation/particletracker.py:
```diff
@@ -1,203 +1,134 @@
-        dt = -self.dt / 2 if init else self.dt
-        b, e = self._interpolate_fields()
+    This ends up causing the magnetic field action to be properly
+    "centered" in time, and the algorithm conserves energy.
```
This is a nice, concise description in the docstring.

namurphy commented on plasmapy/simulation/particletracker.py:
```diff
@@ -1,203 +1,134 @@
-            self.x += self.v * dt
+class ParticleTrackerSolution:
+    """
+    A solution to `ParticleTracker`'s trajectory integration.
```
I like having a separate class for the particle tracker solution.

StanczakDominik replied:
Picked it up from scipy solve ivp and pymc3 outputs. Pymc3 is actually nice in that Hamiltonian Monte Carlo is actually a particle simulation with wacky momenta in model parameter space, so there's plenty tying it back to usual particle stuff and nonlinear dynamics. It's absolutely lovely.

namurphy commented on plasmapy/simulation/particletracker.py:
```diff
@@ -1,203 +1,134 @@
-    Object representing a species of particles: ions, electrons, or simply
-    a group of particles with a particular initial velocity distribution.
+@numba.njit(parallel = True)
+def _boris_push(x, v, b, e, hqmdt, dt):
```
Perhaps de-abbreviate the variable name `hqmdt`?  Maybe something along the lines of `half_q_m_dt`, but there may be something better.

StanczakDominik replied:
I actually think I'll just change it to q, m and take the performance hit of dividing two floats with numba on each step... :D and then if it's a private variable it needs no long name but a comment: # half charge /m mass timestep.

StanczakDominik commented:
I replaced Mayavi with Pyvista, which is looking much healthier as a package, and ta-daa, https://mybinder.org/v2/gh/StanczakDominik/PlasmaPy/RefactorTracker?filepath=notebooks%2FTokamakParticles.ipynb ! I still have some tinkering and generalization to do with the visualize part, but at least it's going to show live. I also think I should now be able to get it to build as a sphinx-gallery example.

StanczakDominik commented:
- [ ] Keep notebooks in examples dir, no plot_ in name so it doesn’t get computed on CI, store pre-rendered notebooks?
- [ ] move coils to plasmapy.plasma.fields

namurphy commented on docs/conf.py:
```diff
@@ -44,8 +61,9 @@
-    'astropy': ('http://docs.astropy.org/en/stable/', None)}
+    'pandas': ('https://pandas.pydata.org/pandas-docs/stable/', None),
+    'astropy': ('http://docs.astropy.org/en/stable/', None),
+    'numba': ('https://numba.pydata.org/numba-doc/dev/', None)}
```
Oh!  This is where these things are defined!

StanczakDominik replied:
Yup :) I'm sorry that this pr does all these little changes in irrelevant places - the joy of refactoring caught me.

namurphy replied:
No worries!  Don't tell anyone that I'm doing that a bit in #728 too!  :scream_cat:  I don't mind us adding on little fixes here and there, since otherwise these little fixes might not get done.  And this time it had the side benefit that I got to learn a little more about the code!

namurphy commented on plasmapy/classes/sources/analyticalfields.py:
```diff
@@ -0,0 +1,47 @@
+        Initialize plasma paramters.
+        The most basic description is composition (ion), temperature,
+        density, and ionization.
+        """
```
The plasma parameters in the docstring don't match the arguments to `__init__` here.

StanczakDominik replied:
Oh yeah, it's a bad habit of mine that I don't update docs as I go. We should really find a way to incorporate pydocstyle into CI, or maybe I should finally learn pre-commit...

StanczakDominik commented:
Okay. Phew. This is mostly done; inspect how it works at https://mybinder.org/v2/gh/StanczakDominik/PlasmaPy/RefactorTracker?filepath=notebooks%2FTokamakParticles.ipynb. I'm completely giving up on adding these to sphinx_gallery for now.

namurphy commented on plasmapy/classes/sources/coils.py:
```diff
@@ -0,0 +1,166 @@
+        match = "interpolate_B" in kwargs.keys()
+        return match
+
+    def visualize(self, figure=None):  # coverage: ignore
```
```suggestion
    def visualize(self, figure=None):
```
My preference would be to keep counting the lines for visualization methods when checking test coverage.  Our coverage percentage will be lower, but it's useful to be able to find where we have untested code.  

An alternative would be to do a regression test and make sure that the hash of the plot object stays the same.  I think this is a strategy that SunPy has used.  The problem is that this seems to be a fragile test, since the hash could change due to upstream changes.  I'm not really sure on the best way to deal with testing plots...

StanczakDominik replied:
I initially did it like this as I had issues getting PyVista to run in CI environments. I've since figured out some tricks to do that.

I'll remove the coverage ignores, make this thread a separate issue, and then in follow-up PR I'll figure out a way to CI plot output (I've been meaning to figure that out anyway) and get this coverage back up as possible. Does that sound good?

namurphy commented on plasmapy/classes/sources/coils.py:
```diff
@@ -0,0 +1,166 @@
+            major radius of the single coil representing the plasma
+        main_current : u.A
+            current through the single plasma-representing coil
+        coil_currents :  Iterable[u.A]
```
```suggestion
        coil_currents : Iterable[u.A]
```

namurphy commented on plasmapy/classes/sources/coils.py:
```diff
@@ -0,0 +1,166 @@
+        main_current : u.A
+            current through the single plasma-representing coil
+        coil_currents :  Iterable[u.A]
+            list of currents through the coils. By default, 8 15-MA coils are created.
```
```suggestion
            list of currents through the coils. By default, eight 15-MA coils are created.
```
For readability...maybe `15 MA` instead of `15-MA` too, unless using the hyphen is standard format.

StanczakDominik replied:
I edited your suggestions to remove the hyphen. Laziness overload.

namurphy commented on plasmapy/examples/plot_particle_stepper.py:
```diff
@@ -2,78 +2,113 @@
-      f"theoretical E0/B0 = {E0/B0:.4f}")
+vmean = solution.velocity.sel(dimension="z").mean().item()
+print(
+    f"The calculated drift velocity is {vmean:.4f} to compare with the "
```
Huh!  I didn't know we could do something like f"{vmean:.4f}".  Format codes remind me of the good ole days of Fortran. :smile_cat:

StanczakDominik replied:
You might like https://pyformat.info/ then :)

namurphy commented on plasmapy/formulary/magnetostatics.py:
```diff
@@ -458,28 +492,92 @@ def magnetic_field(self, p) -> u.T:
-            + np.matmul(np.expand_dims(self.axis_y, 1), np.expand_dims(np.cos(t), 0))
-        )  # (3, n)
+        field = np.zeros((p.shape[0], pt.shape[1], p.shape[1]))
+        for i in numba.prange(pt.shape[1]):  # 300
```
I'm not sure what the `# 300` refers to.

StanczakDominik replied:
The Spartans, obviously! Nah, it's spilled debug juice from when I was trying to rewrite it in Numba. I'll remove it - thanks for pointing it out :)

namurphy commented on plasmapy/simulation/particle_integrators.py:
```diff
@@ -0,0 +1,131 @@
+from astropy import constants
+
+
+@numba.njit(parallel=True)
```
Being a solar physicist (at least nominally), seeing `njit` always makes me think of [NJIT](https://www.njit.edu/), which has a really good solar physics grad program.  More relevantly, it's nice to see you making such good use of Numba!

StanczakDominik replied:
Oh, in New Jersey? I've heard everything's legal there.

namurphy commented on plasmapy/simulation/particle_integrators.py:
```diff
@@ -0,0 +1,131 @@
+
+
+@numba.njit()
+def gamma_from_velocity(velocity):
```
Add a docstring or make it into a private function?

StanczakDominik replied:
Private it is!

namurphy commented on plasmapy/simulation/particle_integrators.py:
```diff
@@ -0,0 +1,131 @@
+
+
+@numba.njit()
+def gamma_from_u(u):
```
Add a docstring or make it into a private function?

namurphy commented on plasmapy/simulation/particle_integrators.py:
```diff
@@ -0,0 +1,131 @@
+
+    References
+    ----------
+    .. [1] Seiji Zenitani and Takayuki Umeda,
+           On the Boris solver in particle-in-cell simulation
+           Physics of Plasmas 25, 112110 (2018); https://doi.org/10.1063/1.5051077
```
Oh yeah...I remember this paper!  Glad you're putting it to good use!

StanczakDominik commented on plasmapy/classes/sources/analyticalfields.py:
```diff
@@ -0,0 +1,78 @@
+
+    def __init__(self, magnetic_field: Callable, electric_field: Callable):
+        test_array = np.zeros((4, 3))
+        if magnetic_field(test_array).shape != test_array.shape:
+            self._interpolate_B = lambda r: np.array([magnetic_field(ri) for ri in r])
+        else:
+            self._interpolate_B = magnetic_field
+        if electric_field(test_array).shape != test_array.shape:
+            self._interpolate_E = lambda r: np.array([electric_field(ri) for ri in r])
+        else:
+            self._interpolate_E = electric_field
```
Sure, except I'll have to think about how best to do that, since I don't want to run this check on every call... I'll figure something out.

StanczakDominik commented on plasmapy/classes/sources/analyticalfields.py:
```diff
@@ -0,0 +1,78 @@
+        return u.Quantity(self._interpolate_E(r), _volt_over_meter, copy=False)
+
+    @classmethod
+    def is_datasource_for(cls, **kwargs):
```
Ah, but that'd imply an understanding of the metaclass mechanic in our `Plasma` class that I don't currently have! I'll cut this for now and add it back when reworking it.

StanczakDominik commented:
Fixes incorporated @namurphy :)

StanczakDominik commented on plasmapy/simulation/particletracker.py:
```diff
@@ -1,215 +1,64 @@
-        r"""
-        Runs a simulation instance.
+        array : xarray.DataArray
+            Array to calculate vector norm of.
```
I was thinking along the lines of "This returns the vector norm **of** this array"...

StanczakDominik commented:
Astropy.units might work better with numba now, so some refactoring and simplification should be possible -> back to drafting board!

StanczakDominik commented:
https://github.com/numba/numba/issues/4637#issuecomment-651422715 would be the thing to try out when coming back to this PR.

However...

I think this has become overly bloated and basically unmergeable. I think the way to go may be to close this one down and break it up in a few smaller, separate PRs, redoing the changes. I still think it was worthwhile, because I now know what to do for this.

@rocco8773, whatcha think?

rocco8773 commented:
@StanczakDominik I think that's a good idea. I agree with all of that.

StanczakDominik commented:
RIP this one, then!
